### PR TITLE
Fix: Request Streaming back pressure

### LINKED ISSUE #1187: Streaming request body stalls, if consumer is too slow
**Describe the bug**

When the request's `bodyAsStream` is run to a slower sink, the whole process stalls and no response is generated. Adding buffering `bodyAsStream.bufferUnlimited` solves the issue, but increases memory consumption and turns off backpressure.

**To Reproduce**
Steps to reproduce the behaviour:

Try the following application: 

```scala
object Main extends ZIOAppDefault {

  private val app = Http.collectZIO[Request] {
    case req @ Method.POST -> !! / "store" => ZIO.scoped (
      for {
        byteCount <- req.bodyAsStream.run(ZSink.fromOutputStreamScoped(
          ZIO.attemptBlockingIO(new BufferedOutputStream(new FileOutputStream("target/test")))
        ))

      } yield Response.text(s"$byteCount stored")
    )
  }

  override def run =
    Server.start(8090, app).exitCode
}
```

Then post some larger file to the server `curl -X POST http://localhost:8090/store --data-binary @somelargerfile` 

The request will be correctly processed for small files, but for larger files the response is never received.

**Expected behaviour**

The request should be processed and a response should be generated. If the server cannot process the data fast enough, it should slow down receiving the data from the client.  (Or, if not possible, at least an error response should be generated)

**Desktop (please complete the following information):**
 - OS: Linux Manjaro, Linux Ubuntu 20.04 , Windows 11
 - Version: 2.0.0-RC6 with ZIO 2.0.0-RC4 (also 2.0.0-RC5 with ZIO 2.0.0-RC3)

### PR DESCRIPTION
**Pull request by gciuloaica**
Fixes #1187

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/HttpData.scala [hunk 1]
```diff
@@ -139,6 +139,21 @@ object HttpData {
...
+      _       <- UIO(
+        unsafeRun { ch =>
+          runtime.unsafeRun(promise.succeed(ch))
+          msg => {
+            runtime.unsafeRun(queue.offer(msg)): Unit
+          }
+        },
+      )
+      ch      <- promise.await
+    } yield queue.mapM(msg => UIO(ch.read()).unless(msg.isLast).as(msg))
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/HttpData.scala]: These needs unit tests.

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 2]
```diff
@@ -1,17 +1,37 @@
...
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Promise
 
 object HttpDataSpec extends DefaultRunnableSpec {
+
+  final case class TestUnsafeReadableChannel(promise: Promise[Unit]) extends UnsafeReadableChannel {
+    override def read(): Unit = promise.success(())
+  }
+
+  final case class TestUnsafeReadableChannel2(queue: Queue[Promise[Unit]]) extends UnsafeReadableChannel {
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]: Please use better name here

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 3]
```diff
@@ -41,7 +61,78 @@ object HttpDataSpec extends DefaultRunnableSpec {
...
+              val probe  = HttpData.UnsafeAsync { cb =>
+                val producer = cb(unsafeChannel)
+                producer(
+                  new UnsafeContent(new DefaultHttpContent(Unpooled.copiedBuffer("abc".toArray.map(_.toByte)))),
+                )
+              }
+              val result = probe.toByteBufStream(ByteBufConfig(4)).runCount
+              assertM(result)(equalTo(1L))
+            } @@ timeout(1 seconds) @@ failing,
+            testM("sending 2 msg should finish the processing with same number of messages.") {
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]:
```suggestion
            testM("process all messages") {
```

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 4]
```diff
@@ -41,7 +61,78 @@ object HttpDataSpec extends DefaultRunnableSpec {
...
+
+              val probe  = HttpData.UnsafeAsync { cb =>
+                val producer = cb(unsafeChannel)
+
+                producer(new UnsafeContent(new DefaultLastHttpContent(Unpooled.EMPTY_BUFFER)))
+              }
+              val result = probe.toByteBufStream(ByteBufConfig(4)).runCount
+              assertM(result)(equalTo(1L))
+            },
+            testM("Sending a chunk without last msg the stream will not complete") {
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]:
```suggestion
            testM("wait for last message") {
```

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 5]
```diff
@@ -41,7 +61,78 @@ object HttpDataSpec extends DefaultRunnableSpec {
               assertM(res)(equalTo("abc\nfoo"))
             },
           ),
+          suite("UnsafeAsync")(
+            testM("sending last msg should finish the processing with 1 msg.") {
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]:
```suggestion
            testM("finish on last message") {
```

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 6]
```diff
@@ -41,7 +61,78 @@ object HttpDataSpec extends DefaultRunnableSpec {
...
+
+              val p1            = Promise[Unit]()
+              val p2            = Promise[Unit]()
+              val queue         = Queue(p1, p2)
+              val unsafeChannel = TestUnsafeReadableChannel2(queue)
+
+              val probe  = HttpData.UnsafeAsync { cb =>
+                val producer = cb(unsafeChannel)
+                producer(
+                  new UnsafeContent(new DefaultHttpContent(Unpooled.copiedBuffer("abc".toArray.map(_.toByte)))),
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]: could we please simplify this test too much code.

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala (hunk 2 above)
Comment (gciuloaica) [reply zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]: done

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala (hunk 6 above)
Comment (gciuloaica) [reply zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]: re-implemented using generators

File: zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala [hunk 7]
```diff
@@ -41,7 +62,88 @@ object HttpDataSpec extends DefaultRunnableSpec {
...
+                )
+              }
+              val result = probe.toByteBufStream(ByteBufConfig(4)).runCount
+              assertM(result)(equalTo(1L))
+            } @@ timeout(1 seconds) @@ failing,
+            testM("process all messages.") {
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]: this and the next test are doing the same thing

Comment (gciuloaica) [reply zio-http/src/test/scala/zhttp/http/HttpDataSpec.scala]: done

File: zio-http/src/main/scala/zhttp/http/HttpData.scala (hunk 1 above)
Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/http/HttpData.scala]: Done

**Comment by tusharmath** [comment]
I tried this PR with this test code — 


```scala
  val app = Http.collectZIO[Request] { case req @ Method.PUT -> !! / "echo" =>
    for {
      byteCount <- req.bodyAsStream.run(
        ZSink.fromOutputStream(
          new BufferedOutputStream(new FileOutputStream("target/test")),
        ),
      )

    } yield Response.text(s"$byteCount stored")
  }
```

The source file uploaded and the final file that's created had different checksums @gciuloaica we might have to debug this further I think.
I tried it with a 1GB file.

**Comment by gciuloaica** [comment]
> 

@tusharmath based on some debug, the issue is not in zio-http but in zio I guess:

`ZStream
      .fromFile(Paths.get("/Users/gciuloaica/1g.img"))
      .run(ZSink.fromOutputStream(new BufferedOutputStream(new FileOutputStream("target/test"))))`
      
This code will generate a file with a different size than original and implicitly a different hash.

**Comment by gciuloaica** [comment]
> I tried this PR with this test code —
> 
> ```scala
>   val app = Http.collectZIO[Request] { case req @ Method.PUT -> !! / "echo" =>
>     for {
>       byteCount <- req.bodyAsStream.run(
>         ZSink.fromOutputStream(
>           new BufferedOutputStream(new FileOutputStream("target/test")),
>         ),
>       )
> 
>     } yield Response.text(s"$byteCount stored")
>   }
> ```
> 
> The source file uploaded and the final file that's created had different checksums @gciuloaica we might have to debug this further I think. I tried it with a 1GB file.

@tusharmath the issue is generated by usage of `BufferedOutputStream`. The stream is not getting closed and the data from internal buffer is not getting flushed. With a small change like:

```scala
 val app = Http.collectZIO[Request] { case req @ Method.PUT -> !! / "echo" =>
    val out = new BufferedOutputStream(new FileOutputStream("target/test"))
    for {
      byteCount <- req.bodyAsStream.run(
        ZSink.fromOutputStream(out),
      )
      _         <- UIO(out.close())

    } yield Response.text(s"$byteCount stored")
  }
``` 
it will fix the issue. However it should be preferable to use `ZSink.fromFile` instead that will handle also closing of all open streams.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**